### PR TITLE
Add support for AWS metadata (tags)

### PR DESCRIPTION
**Pull request by 0x2b3bfa0**
Part of https://github.com/iterative/terraform-provider-iterative/issues/185; keeps a default [`Name` tag](https://docs.aws.amazon.com/whitepapers/latest/tagging-best-practices/standardize-names-for-aws-resources.html) for backwards compatibility, but [perhaps shouldn't](https://github.com/iterative/terraform-provider-iterative/pull/187#discussion_r702439138).

### CODE REVIEW

File: iterative/resource_machine.go [hunk 1]
```diff
@@ -122,6 +122,14 @@ func machineSchema() *map[string]*schema.Schema {
...
 			Default:  "",
 		},
+		"metadata": &schema.Schema{
+			Type:     schema.TypeMap,
+			ForceNew: true,
+			Optional: true,
+			Elem: &schema.Schema{
+				Type: schema.TypeString,
+			},
+		},
```
Comment (0x2b3bfa0) [inline iterative/resource_machine.go:132]:
```suggestion
		"metadata": &schema.Schema{
			Type:     schema.TypeMap,
			ForceNew: true,
			Optional: true,
			Elem: &schema.Schema{
				Type: schema.TypeString,
			},
		},
```

Removing in favor of https://github.com/iterative/terraform-provider-iterative/pull/193

File: iterative/resource_runner.go [hunk 2]
```diff
@@ -153,6 +153,14 @@ func resourceRunner() *schema.Resource {
...
 				Default:  "",
 			},
+			"metadata": &schema.Schema{
+				Type:     schema.TypeMap,
+				ForceNew: true,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
```
Comment (0x2b3bfa0) [inline iterative/resource_runner.go:163]:
```suggestion
			"metadata": &schema.Schema{
				Type:     schema.TypeMap,
				ForceNew: true,
				Optional: true,
				Elem: &schema.Schema{
					Type: schema.TypeString,
				},
			},
```
Removing in favor of #193

File: iterative/aws/provider.go [hunk 3]
```diff
@@ -32,6 +31,15 @@ func ResourceMachineCreate(ctx context.Context, d *schema.ResourceData, m interf
 	securityGroup := d.Get("aws_security_group").(string)
 	spot := d.Get("spot").(bool)
 	spotPrice := d.Get("spot_price").(float64)
+
+	metadata := map[string]string{
+		"Name": d.Get("name").(string),
+		"Id":   d.Id(),
```
Comment (0x2b3bfa0) [inline iterative/aws/provider.go:37]: Should we really provide any defaults?

Comment (0x2b3bfa0) [reply iterative/aws/provider.go:37]: The  [`Name` tag](https://docs.aws.amazon.com/whitepapers/latest/tagging-best-practices/standardize-names-for-aws-resources.html) is being used by AWS to name instances in the console, but the `Id` tag doesn't name any sense and should probably be removed unless users ask for it.

Comment (DavidGOrtega) [reply iterative/aws/provider.go:37]: Let me review it but its the way to relate the runner and the machine, as far as I remember

Comment (DavidGOrtega) [reply iterative/aws/provider.go:37]: Id and Name are different, not saying that we should not unify them or check. As far as I remember we also had an open issue to remove name in the runner and just stay with tags

Comment (0x2b3bfa0) [reply iterative/aws/provider.go:37]: https://github.com/iterative/cml/issues/525

Comment (0x2b3bfa0) [reply iterative/aws/provider.go:37]: Then, let's leave this as–is, fully backwards compatible, and then address https://github.com/iterative/cml/issues/525 separately.

**Comment by DavidGOrtega** [comment]
Crashes for me with:

```tf
resource "iterative_cml_runner" "runner-gh" {
    token = "XXXX"
    repo = "XXXX"
    driver = "gitlab"
    labels = "aws"
    idle_timeout = 0

    cloud = "aws"
    region = "us-west"
    instance_type = "t2.micro"
    spot = true
    ssh_private = "XXXX"
}
```

Also does not allow me to use the expected metatags

```tf
resource "iterative_cml_runner" "runner-gh" {
   metadata = {
      project = "example"
      creator = "anybody"
    }

    token = "XXXX"
    repo = "XXXX"
    driver = "gitlab"
    labels = "aws"
    idle_timeout = 0

    cloud = "aws"
    region = "us-west"
    instance_type = "t2.micro"
    spot = true
    ssh_private = "XXXX"
}
```

```sh

Error: Unsupported argument

  on main.tf line 98, in resource "iterative_cml_runner" "runner-gh-5":
  98:     metadata = {

An argument named "metadata" is not expected here.
```

**Review by DavidGOrtega — CHANGES_REQUESTED** [review]
Crashes for me with:

```tf
resource "iterative_cml_runner" "runner-gh" {
    token = "XXXX"
    repo = "XXXX"
    driver = "gitlab"
    labels = "aws"
    idle_timeout = 0

    cloud = "aws"
    region = "us-west"
    instance_type = "t2.micro"
    spot = true
    ssh_private = "XXXX"
}
```

Also does not allow me to use the expected metatags

```tf
resource "iterative_cml_runner" "runner-gh" {
   metadata = {
      project = "example"
      creator = "anybody"
    }

    token = "XXXX"
    repo = "XXXX"
    driver = "gitlab"
    labels = "aws"
    idle_timeout = 0

    cloud = "aws"
    region = "us-west"
    instance_type = "t2.micro"
    spot = true
    ssh_private = "XXXX"
}
```

```sh

Error: Unsupported argument

  on main.tf line 98, in resource "iterative_cml_runner" "runner-gh":
  98:     metadata = {

An argument named "metadata" is not expected here.
```

**Comment by 0x2b3bfa0** [comment]
@DavidGOrtega, can you please kindly check if you've applied these changes over `master` @  https://github.com/iterative/terraform-provider-iterative/commit/4044aee23ea8209ba5244faeaafa8fdbfe92a26e ?

**Comment by 0x2b3bfa0** [comment]
It looks like you don't have the changes from https://github.com/iterative/terraform-provider-iterative/pull/193 in your working tree.

**Comment by DavidGOrtega** [comment]
> It looks like you don't have the changes from #193 in your working tree.

Ok I pull origin and started to work, however I came up with this error

```sh
Error: Failed creating the machine: Not able to decode: operation error EC2: 
RequestSpotInstances, https response error StatusCode: 400, 
RequestID: 16f4fa79-07f2-4e8a-9446-2543e2804c9c, api error InvalidParameter: 'spot-instance-request' is not a valid taggable resource type for this operation.
```

File: iterative/aws/provider.go [hunk 4]
```diff
@@ -225,7 +237,8 @@ func ResourceMachineCreate(ctx context.Context, d *schema.ResourceData, m interf
 				SubnetId:            aws.String(subnetID),
 				BlockDeviceMappings: blockDeviceMappings,
 			},
-			InstanceCount: aws.Int32(1),
+			InstanceCount:     aws.Int32(1),
+			TagSpecifications: resourceTagSpecifications("spot-instance-request", metadata),
```
Comment (0x2b3bfa0) [inline iterative/aws/provider.go:241]:
```suggestion
			TagSpecifications: resourceTagSpecifications("spot-instances-request", metadata),
```

**Comment by 0x2b3bfa0** [comment]
Sorry, this time it was my fault. 🙃  Please pull & retry.

**Comment by 0x2b3bfa0** [comment]
> _I do not think last time was my fault either_ 😁

Probably not / not completely. I should have merged #193 back into the child branches as soon as it was merged.

**Comment by DavidGOrtega** [comment]
Name is not created and after an error (with my runner credentials) the machine was not removed. Have been this tested by you? I mean, the PR is ready to be tested?

image

**Comment by DavidGOrtega** [comment]
Destroy does not work either

**Comment by 0x2b3bfa0** [comment]
> _Name is not created and after an error (with my runner credentials) the machine was not removed._ [...] _Destroy does not work either._

This sounds like a failure creating tags... 🤔  Destroy can't work without the `Id` tag either.

> _Have been this tested by you? I mean, the PR is ready to be tested?_

I tested it before separating #183 from #187, and again just now, and it worked like a charm. My typo earlier was a keyboard/brain slip when splitting the pull request in two.

**Comment by DavidGOrtega** [comment]
🤔 I did not have any null error. It said machine destroyed succesfully

**Comment by 0x2b3bfa0** [comment]
Well, I did, but it was just because I had expired sandbox credentials. 🙃  Perhaps I have discovered #200 accidentally. 🐈 ⌨️

**Comment by 0x2b3bfa0** [comment]
Both `apply` and `destroy` work for me as expected. 🤔 🚂  Can you please check you're in https://github.com/iterative/terraform-provider-iterative/pull/187/commits/ee283951a28f7e629aa65489a0ed92781895ff98 ?

**Comment by DavidGOrtega** [comment]
Same, instance without name and still running while it should have been destroyed by the TPI

image

**Comment by 0x2b3bfa0** [comment]
Can you please post the HCL2 file used to generate that instance? I'd like to try it locally.

**Comment by DavidGOrtega** [comment]
> Can you please post the HCL2 file used to generate that instance? I'd like to try it locally.

Which file? Ii do use a tf file not such HCL2

**Comment by 0x2b3bfa0** [comment]
@DavidGOrtega, I forgot that spot instance request tags don't get propagated to instances; that only happens for fleets using launch templates. Now it should work as expected. 🤦🏼‍♂️ 😅

**Comment by 0x2b3bfa0** [comment]
> _Which file? Ii do use a tf file not such HCL2_

Yes, I meant the `.tf` file, which is written in [HCL[2]](https://github.com/hashicorp/hcl), as opposed to `.tf.json` files that are written in JSON.

**Comment by DavidGOrtega** [comment]
@0x2b3bfa0 

Now seems to work. To sure how it worked for you before.

Is this PR still needing https://github.com/iterative/terraform-provider-iterative/pull/183? If not why https://github.com/iterative/terraform-provider-iterative/pull/183?

**Comment by 0x2b3bfa0** [comment]
> Now seems to work. To sure how it worked for you before.

I didn't test spot instances. 🐒

**Comment by 0x2b3bfa0** [comment]
> _Is this PR still needing #183? If not why #183?_

Not anymore, but it doesn't hurt either; you save a few pointers and there isn't any major change for our use case.

**Review by DavidGOrtega — APPROVED** [review]
🚀 lgtm

**Pull request merged**